### DISCUSSION TITLE
Wire visualization in surface. Example for SlotW31B

### DISCUSSION THREAD

**Discussion by zmladen** [Eomys/pyleecan#714] [General]
I'm not sure if I should open a new topic for this or if it should be part of the discussion for SlotW31B. I've made some progress on the wire visualization inside active surfaces. Here's the algorithm I've come up with for now:

1. When creating the SurfLine for the active area, set reference segments (ref_seg). These represent the lines where the initial wire layer will be placed. For that, I had to create a new class, SegmentRef, with an is_ref property.
2. After dividing the active area into subsurfaces, for each one, find the previous segment. Or next if (is_internal=True)
3. Move the segments by a distance (half of the wire diameter).
4. Find the intersection points. These represent the initial placement of the first wire (conductor).
5. Create a grid of points, resulting in a triangular packing of circles.
6. Filter the points that lie inside the polygon.

Any other ideas?

![image](https://github.com/user-attachments/assets/f402bbbc-2308-4d0a-8bc9-f60afb73ac4a)

![image](https://github.com/user-attachments/assets/8e4cff51-67a7-40c6-b324-ee5b722b98dc)

![image](https://github.com/user-attachments/assets/600034c7-ffff-45c7-a2ff-bf4747ed8161)

![image](https://github.com/user-attachments/assets/97d4a1c2-6db8-4484-98c6-38d75d431d97)

**Comment by zmladen** [comment]
I implemented the algorithm in the SurfLine object as suggested by Pierre. To support this, I added a couple of new functions (see the image below). Additionally, if no reference line is set when defining the active area, the get_ref_line function will instead identify the longest line.

![image](https://github.com/user-attachments/assets/288a132c-d7da-4b8a-a7ef-565853843e56)

To plot the results currently i added "is_show_wires" in the plot_active method of the "Slot" class and the following code:

`
    if is_show_wires:

        circles = []

        for surface in surf_list:
            w_cord = surface._get_wire_coordinate(self.get_Wwire(), self.is_outwards())

            for jj in range(Zs):

                for wire in w_cord:
                    rotated_wire = wire * exp(1j * 2 * pi * jj / Zs)

                    circle = Circle(
                        (rotated_wire.real, rotated_wire.imag),
                        self.get_Wwire() / 2,
                    )
                    circles.append(circle)

        # Use PatchCollection to improve performance
        wire_collection = PatchCollection(
            circles, facecolor="none", edgecolor="black", zorder=2
        )

        ax.add_collection(wire_collection)
`
This is the resulting plot.

![image](https://github.com/user-attachments/assets/63e615f6-a006-4d55-aa63-b405be196a36)

 It works fine for the slot I have created but others need to be tested as well. I am not sure how to procede further. Shoud I create separate, e.g. plot_wires funciton in Slot to focus the results to only one slot? I would like to plot the conductor directly instead of round wires as conductor could have multiply wires as well...

**Comment by BonneelP** [comment]
Hello,

That looks amazing :) 

Instead of adding a "is_ref" property maybe you can use the "prop_dict" property of any Line. It is what we use to set "this line is particular" (for instance to set Boundary conditions). 

For the next step, I would see several possibilities:
- What about slots arcs ? (like SlotW11 ?)
- You can investigate calling this method within the build_geometry method of LamSlotWind by adding a new boolean option that would return the circles instead of the surfaces. Ideally you can take Nt_coil and the conductor dimensions directly from the winding and winding.conductor object.
- You can indeed go one step further, take your conductors and add some wires within by adding a new method within the Circle object (for that I think we already have few code lines)

Best regards,
Pierre

**Comment by zmladen** [comment]
Hello Pierre,

Thanks for your feedback! I'll work on implementing your suggestions next week. I still need to come up with a solution for arcs, but it should be doable. Returning conductors instead of active surfaces is a great idea, especially if one needs to consider skin and proximity effects. 

In the mean time i changed a Conductor12 a bit (I added new class Conductor12B). I wanted to be able to use actual conductor objects inside a slot. This enables better visualization for multiwire conducotrs as well. Here first resuts:

![image](https://github.com/user-attachments/assets/b3195665-52d8-4e3e-be38-0d124c326371)

As i have a need for more thant 4 wires in conductor i have extended the plot method of Conductor12B. It is ok for now but I think we can do better. We might need separate discusson for that as it is not trivial.

![image](https://github.com/user-attachments/assets/c7e29d1c-1786-4607-956b-64bde3d6c0f2)

![image](https://github.com/user-attachments/assets/4c724abf-7198-4995-b777-ac6a9d4ca437)

I used this approach just to test, but I kept your solution up to 4 wires.
[https://math.stackexchange.com/questions/1283085/hexagon-packing-in-a-circle](url)

How can i add a new color to config_dict? I would like to define new color when plotting only the slot. Currently the blue color is used for both, slot and conductor isolation which does not look very nice.

`
SLOT_COLOR = config_dict["PLOT"]["COLOR_DICT"]["SLOT_COLOR"]
`
Regards,
Mladen

**Reply by BonneelP** [reply]
Very nice :) 
The plot method of Conductor12  was a placeholder for documentation by waiting to implement what you did ;)
The website architecture would need to be improved but just to be sure we have few elements on the winding vocabulary here: https://www.pyleecan.org/winding.convention.html
The point to return "conductor surfaces" instead of "active surface" is that it will enable the definition in the FEA software more easily. We will need to set a proper unique label to each conductor surface. Then from the label, the FEA software will know which property to set in which surface. The idea of providing a unique label is to enable setting different property in any of the surface (for instance we could simulate a broken wire or exotic topology).
Sure you can add new colors to the config_dict :)

**Reply by zmladen** [reply]
I have added new color to "pyleecan_color.json", e.g. "SLOT_COLOR": "mangenta". But when i try to import it in plot.py i get en error:

`
  File "C:\Work\Python\pyleecan\pyleecan\Methods\Slot\SlotW31B\plot.py", line 11, in <module>
    SLOT_COLOR = config_dict["PLOT"]["COLOR_DICT"]["SLOT_COLOR"]
KeyError: 'SLOT_COLOR'
(venv39) PS C:\Work\Python\pyleecan> 
`

**Comment by zmladen** [comment]
Hello Pierre,

following to your suggestions. I added the followig:

1. Conductor has now "build_geometry" method which creates geoemtry of single wire
2. SurfaLine has now "build_geometry_wire" which generates the wire geometry for each active surface. It also has method for calculating conductor centers based on the conductor properties as you suggested.
3. In the build_geometry of the LamSLotWind i added "is_add_wires" parameter that can be used to control if the wires or active surfaces are to be returned. The same parameter is in plot method.

This is added code in build_geometry of LamSLotWind:

`
        for ii in range(Zs // sym):  # for each slot
            # for each part of the winding surface in the slot
            for surf in surf_Wind:
                new_surf = surf.copy()

                if is_add_wires:
                    if self.winding.conductor is not None:
                        for surf_wire in surf.build_geometry_wire(
                            self.winding.conductor
                        ):
                            surf_wire.rotate(ii * slot_pitch)

                            surf_list.append(surf_wire)

                            surf_wire.label = update_RTS_index(
                                label=surf.label, S_id=ii, surf_type_label=WIND_LAB
                            )
                else:
                    new_surf.label = update_RTS_index(
                        label=surf.label, S_id=ii, surf_type_label=WIND_LAB
                    )
                    new_surf.rotate(ii * slot_pitch)

                    surf_list.append(new_surf)
`
The same can be now done for rectangular type conductors as well if needed.

Here the results:

![image](https://github.com/user-attachments/assets/4a23c51a-f730-4e31-9165-d54f20efbbd4)

![image](https://github.com/user-attachments/assets/4e55850d-be36-4bc4-baa0-7d96a925a4ec)

Next steps for me are to check the calculations of the phase resistance. It would be great if you take a look at the error i got when adding new color to "pyleecan_color.json".

Regards,
Mladen
